src/pdm/utils.py

```diff
@@ -453,3 +453,27 @@ def package_installed(package_name: str) -> bool:
         return False
     else:
         return True
+
+
+def validate_project_name(name: str) -> bool:
+    """Check if the project name is valid or not"""
+    import string
+
+    allowed_chars = set(string.ascii_letters + string.digits + "_-")
+    if set(name) - allowed_chars:
+        return False
+
+    if name[0].isdigit():
+        return False
+
+    import keyword
+
+    if keyword.iskeyword(name):
+        return False
+
+    import builtins
+
+    if name in dir(builtins):
+        return False
+
+    return True
```

Comment:
We can only consider the standard of [PEP 426](https://peps.python.org/pep-0426/#name) which even gives a regex pattern that you can use. No need to check the keywords and builtin.